File: cc/019.cc
// Project Euler Problem 019
//
// How many Sundays fell on the first of the month during the
// twentieth century?

#include <cstdio>

// Thanks, Tomohiko Sakamoto
int dow(int y, int m, int d) {
  static int t[] = {0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4};
  y -= m < 3;
  return (y + y/4 - y/100 + y/400 + t[m-1] + d) % 7;
}

int main(char argv[]) {
  int count = 0;
  for (int y = 1901; y < 2001; y++) {
    for (int m = 1; m<=12; m++) {
      count += !(dow(y, m, 1));
    }
  }
  printf("%d\n", count);
}

File: go/problems/019.go
// Project Euler Problem 019
//
// How many Sundays fell on the first of the month during the
// twentieth century?

package problems

import "fmt"

func Problem019() string {
	months := [...]int{31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31}
	sundays := 0
	day := 366 // Monday is 0, 1900 had 366 days
	for year := 1901; year < 2001; year++ {
		isLeap := (year%4 == 0) && ((year%100 != 0) || year%400 == 0)
		for month := 0; month < 12; month++ {
			if day%7 == 6 {
				sundays++
			}
			day += months[month]
			if month == 1 && isLeap {
				day += 1
			}
		}
	}
	return fmt.Sprint(sundays)
}

func init() {
     Functions[19] = Problem019
}
